# Guessing Income  Project
This project is a Data Science Bootcamp project; it is ment to build a modle to guess the income of each indivisual from a dataset contains demographic and financial information about selected people.

In [ ]:
import pandas as pd
import numpy as np

## Exploring the Dataset

Here are the attributes information as they are described in [UCI Machine Learning Repository](https://archive.ics.uci.edu/ml/datasets/adult):

Attribute Information:

- **age:** continuous.
- **workclass:** Private, Self-emp-not-inc, Self-emp-inc, Federal-gov, Local-gov, State-gov, Without-pay, Never-worked.
- **fnlwgt:** continuous.
- **education:** Bachelors, Some-college, 11th, HS-grad, Prof-school, Assoc-acdm, Assoc-voc, 9th, 7th-8th, 12th, Masters, 1st-4th, 10th, Doctorate, 5th-6th, Preschool.
- **education-num:** continuous.
- **marital-status:** Married-civ-spouse, Divorced, Never-married, Separated, Widowed, Married-spouse-absent, Married-AF-spouse.
- **occupation:** Tech-support, Craft-repair, Other-service, Sales, Exec-managerial, Prof-specialty, Handlers-cleaners, Machine-op-inspct, Adm-clerical, Farming-fishing, Transport-moving, Priv-house-serv, Protective-serv, Armed-Forces.
- **relationship:** Wife, Own-child, Husband, Not-in-family, Other-relative, Unmarried.
- **race:** White, Asian-Pac-Islander, Amer-Indian-Eskimo, Other, Black.
- **sex:** Female, Male.
- **capital-gain:** continuous.
- **capital-loss:** continuous.
- **hours-per-week:** continuous.
- **native-country:** United-States, Cambodia, England, Puerto-Rico, Canada, Germany, Outlying-US(Guam-USVI-etc), India, Japan, Greece, South, China, Cuba, Iran, Honduras, Philippines, Italy, Poland, Jamaica, Vietnam, Mexico, Portugal, Ireland, France, Dominican-Republic, Laos, Ecuador, Taiwan, Haiti, Columbia, Hungary, Guatemala, Nicaragua, Scotland, Thailand, Yugoslavia, El-Salvador, Trinadad&Tobago, Peru, Hong, Holand-Netherlands.
- **income:** >50K, <=50K.


In [15]:
data_df = pd.read_csv('../Dataset/adult.data')
data_df.head()

age          workclass  fnlwgt education_level  education-num  \
0   39          State-gov   77516       Bachelors             13   
1   50   Self-emp-not-inc   83311       Bachelors             13   
2   38            Private  215646         HS-grad              9   
3   53            Private  234721            11th              7   
4   28            Private  338409       Bachelors             13   

        marital-status          occupation    relationship    race      sex  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

   capital-gain  capital-loss  hours-per-week  native-country  income  
0          2174             0              40   United-States   <=50K  
1             0             0              13   United-States   <=50K  
2             0             0              40   United-States   <=50K  
3             0             0              40   United-States   <=50K  
4             0             0              40            Cuba   <=50K

In [12]:
data_df.shape

(32561, 15)

In [13]:
data_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   age              32561 non-null  int64 
 1   workclass        32561 non-null  object
 2   fnlwgt           32561 non-null  int64 
 3   education_level  32561 non-null  object
 4   education-num    32561 non-null  int64 
 5   marital-status   32561 non-null  object
 6   occupation       32561 non-null  object
 7   relationship     32561 non-null  object
 8   race             32561 non-null  object
 9   sex              32561 non-null  object
 10  capital-gain     32561 non-null  int64 
 11  capital-loss     32561 non-null  int64 
 12  hours-per-week   32561 non-null  int64 
 13  native-country   32561 non-null  object
 14  income           32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


In [14]:
data_df.describe()

age        fnlwgt  education-num  capital-gain  capital-loss  \
count  32561.000000  3.256100e+04   32561.000000  32561.000000  32561.000000   
mean      38.581647  1.897784e+05      10.080679   1077.648844     87.303830   
std       13.640433  1.055500e+05       2.572720   7385.292085    402.960219   
min       17.000000  1.228500e+04       1.000000      0.000000      0.000000   
25%       28.000000  1.178270e+05       9.000000      0.000000      0.000000   
50%       37.000000  1.783560e+05      10.000000      0.000000      0.000000   
75%       48.000000  2.370510e+05      12.000000      0.000000      0.000000   
max       90.000000  1.484705e+06      16.000000  99999.000000   4356.000000   

       hours-per-week  
count    32561.000000  
mean        40.437456  
std         12.347429  
min          1.000000  
25%         40.000000  
50%         40.000000  
75%         45.000000  
max         99.000000

In [25]:
data_df.duplicated().value_counts()

False    32537
True        24
dtype: int64

In [24]:
data_df[data_df.duplicated()]

age          workclass  fnlwgt education_level  education-num  \
4881    25            Private  308144       Bachelors             13   
5104    90            Private   52386    Some-college             10   
9171    21            Private  250051    Some-college             10   
11631   20            Private  107658    Some-college             10   
13084   25            Private  195994         1st-4th              2   
15059   21            Private  243368       Preschool              1   
17040   46            Private  173243         HS-grad              9   
18555   30            Private  144593         HS-grad              9   
18698   19            Private   97261         HS-grad              9   
21318   19            Private  138153    Some-college             10   
21490   19            Private  146679    Some-college             10   
21875   49            Private   31267         7th-8th              4   
22300   25            Private  195994         1st-4th              2   
22367   44            Private  367749       Bachelors             13   
22494   49   Self-emp-not-inc   43479    Some-college             10   
25872   23            Private  240137         5th-6th              3   
26313   28            Private  274679         Masters             14   
28230   27            Private  255582         HS-grad              9   
28522   42            Private  204235    Some-college             10   
28846   39            Private   30916         HS-grad              9   
29157   38            Private  207202         HS-grad              9   
30845   46            Private  133616    Some-college             10   
31993   19            Private  251579    Some-college             10   
32404   35            Private  379959         HS-grad              9   

            marital-status          occupation    relationship  \
4881         Never-married        Craft-repair   Not-in-family   
5104         Never-married       Other-service   Not-in-family   
9171         Never-married      Prof-specialty       Own-child   
11631        Never-married        Tech-support   Not-in-family   
13084        Never-married     Priv-house-serv   Not-in-family   
15059        Never-married     Farming-fishing   Not-in-family   
17040   Married-civ-spouse        Craft-repair         Husband   
18555        Never-married       Other-service   Not-in-family   
18698        Never-married     Farming-fishing   Not-in-family   
21318        Never-married        Adm-clerical       Own-child   
21490        Never-married     Exec-managerial       Own-child   
21875   Married-civ-spouse        Craft-repair         Husband   
22300        Never-married     Priv-house-serv   Not-in-family   
22367        Never-married      Prof-specialty   Not-in-family   
22494   Married-civ-spouse        Craft-repair         Husband   
25872        Never-married   Handlers-cleaners   Not-in-family   
26313        Never-married      Prof-specialty   Not-in-family   
28230        Never-married   Machine-op-inspct   Not-in-family   
28522   Married-civ-spouse      Prof-specialty         Husband   
28846   Married-civ-spouse        Craft-repair         Husband   
29157   Married-civ-spouse   Machine-op-inspct         Husband   
30845             Divorced        Adm-clerical       Unmarried   
31993        Never-married       Other-service       Own-child   
32404             Divorced       Other-service   Not-in-family   

                      race      sex  capital-gain  capital-loss  \
4881                 White     Male             0             0   
5104    Asian-Pac-Islander     Male             0             0   
9171                 White   Female             0             0   
11631                White   Female             0             0   
13084                White   Female             0             0   
15059                White     Male             0             0   
17040                White     Male             0             0   
18555              

In [49]:
def check_unique(df):
    for i in df.columns:
        print(f'There are {len(df[i].unique())} unique {i}')

In [50]:
check_unique(data_df)

There are 73 unique age
There are 9 unique workclass
There are 21648 unique fnlwgt
There are 16 unique education_level
There are 16 unique education-num
There are 7 unique marital-status
There are 15 unique occupation
There are 6 unique relationship
There are 5 unique race
There are 2 unique sex
There are 119 unique capital-gain
There are 92 unique capital-loss
There are 94 unique hours-per-week
There are 42 unique native-country
There are 2 unique income


In [121]:
def find_question_mark(df):
    sums=0
    for i in df.columns:
        for j, jj in zip(df[i].value_counts().index, df[i].value_counts()):
            if (j == ' ?'):
                print(f'The column {i} has {jj} "?" values.')
                sums+=jj
    print(f'The total of "?" values is {sums}')

In [122]:
find_question_mark(data_df)

The column workclass has 1836 "?" values.
The column occupation has 1843 "?" values.
The column native-country has 583 "?" values.
The total of "?" values is 4262


In [150]:
def find_empty_space(df):
    space = ' '
    sums=0
    for i in df.columns:
        for j, jj in zip(df[i].value_counts().index, df[i].value_counts()):
            if(space in str(j)):
                sums+=jj
    print(f'The total of " " values is {sums/9} in each column of the str type columns.')
    
#Here we have only 9 columns with string type values

In [151]:
find_empty_space(data_df)

The total of " " values is 32561.0 in each column of the str type columns.


### Exoloring Summary

1. The dataset consists of 15 columns and 32561 rows.
2. We have 32561 non_null records in all of the 15 columns in the dataset. 
3. The data types are str and int data types. 
4. From the description of the dataset, I found the following:
    - The ages are between 90 and 17 years old.
    - The Capitl gain is between 0 and 99999. 
    - The Capital loss is between 0 and 4356. 
    - The Working hours per week are between 1 hour to 99 hours.
5. By checking the duplicated records, I found that there are no duplication as the rows overall are unique. 
6. The income information is classified as it follows:
    - \>50K yearlly income.
    - <=50K yearlly income.

### Cleaning Data Findings

**Over All:**
- **Finding 1:**
    - The columns have (-) to sperate between words.
- **Action:**
    - The columns will be renamed with (_) as a sperater between words. 
<br>
<br>

**Values issues Findings:**
- **Finding 1:**
    - The following columns has 4262 records as "?" value:
        - workclass
        - occupation
        - native-country
- **Action:**
    - Remove the records as we don't have the missing information to fill them. 
- **Finding 2:**
    - An extra space at the begaining of each string valuein the dataset.
- **Action:**
    - Remove the extra space.
- **Finding 3:**
    - Column "fnlwgt" has no meaning that may help in this project.
- **Action:**
    - Drop this column.

## Cleaning Data

**Fixing the column headers format**

In [396]:
#Taking copy from the original dataframe for the cleaning process.
cpy_data_df = data_df

In [397]:
def rename_columns(df):
    for i, idx in zip(df.columns, range(len(df.columns))):
        if('-' in i):
            i=i.replace('-','_')
            df.rename(columns={df.columns[idx]: i}, inplace=True)

In [398]:
rename_columns(cpy_data_df)

In [399]:
cpy_data_df.columns

Index(['age', 'workclass', 'fnlwgt', 'education_level', 'education_num',
       'marital_status', 'occupation', 'relationship', 'race', 'sex',
       'capital_gain', 'capital_loss', 'hours_per_week', 'native_country',
       'income'],
      dtype='object')

**Remove Spaces**

In [400]:
def remove_spaces(df):
    for i in df.columns:
        if(df[i].dtypes == 'object'):
            df[i]=df[i].str.lstrip().values

In [401]:
remove_spaces(cpy_data_df)

Checking if there are still empty spaces

In [402]:
find_empty_space(cpy_data_df)

The total of " " values is 0.0 in each column of the str type columns.


**Removing records contains "?"** <br>
? seems related to "not available or missing" data; because that we don't have the values to replace them with the right data, it is better to drop the rows to not effect the modle's output. 

In [403]:
cpy_data_df=cpy_data_df[cpy_data_df['workclass']!='?']
cpy_data_df=cpy_data_df[cpy_data_df['occupation']!='?']
cpy_data_df=cpy_data_df[cpy_data_df['native_country']!='?']

In [404]:
#def remove_question_mark(df):
#    cols = ['workclass', 'occupation', 'native_country']
#    for i in cols:
#        df=df[df[i]!='?']

In [405]:
#remove_question_mark(cpy_data_df)

Checking if there are still records with "?" values

In [406]:
def find_question_mark(df):
    sums=0
    for i in df.columns:
        for j, jj in zip(df[i].value_counts().index, df[i].value_counts()):
            if (j == '?'):
                print(f'The column {i} has {jj} "?" values.')
                sums+=jj
    print(f'The total of "?" values is {sums}')

In [407]:
find_question_mark(cpy_data_df)

The total of "?" values is 0


**Drop "fnlwgt" Column**

In [408]:
cpy_data_df.drop('fnlwgt',axis=1, inplace=True)
cpy_data_df=cpy_data_df.reset_index(drop=True)

In [413]:
cleaned_data_df=cpy_data_df

In [414]:
cleaned_data_df

age         workclass education_level  education_num  \
0       39         State-gov       Bachelors             13   
1       50  Self-emp-not-inc       Bachelors             13   
2       38           Private         HS-grad              9   
3       53           Private            11th              7   
4       28           Private       Bachelors             13   
...    ...               ...             ...            ...   
30157   27           Private      Assoc-acdm             12   
30158   40           Private         HS-grad              9   
30159   58           Private         HS-grad              9   
30160   22           Private         HS-grad              9   
30161   52      Self-emp-inc         HS-grad              9   

           marital_status         occupation   relationship   race     sex  \
0           Never-married       Adm-clerical  Not-in-family  White    Male   
1      Married-civ-spouse    Exec-managerial        Husband  White    Male   
2                Divorced  Handlers-cleaners  Not-in-family  White    Male   
3      Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4      Married-civ-spouse     Prof-specialty           Wife  Black  Female   
...                   ...                ...            ...    ...     ...   
30157  Married-civ-spouse       Tech-support           Wife  White  Female   
30158  Married-civ-spouse  Machine-op-inspct        Husband  White    Male   
30159             Widowed       Adm-clerical      Unmarried  White  Female   
30160       Never-married       Adm-clerical      Own-child  White    Male   
30161  Married-civ-spouse    Exec-managerial           Wife  White  Female   

       capital_gain  capital_loss  hours_per_week native_country income  
0              2174             0              40  United-States  <=50K  
1                 0             0              13  United-States  <=50K  
2                 0             0              40  United-States  <=50K  
3                 0             0              40  United-States  <=50K  
4                 0             0              40           Cuba  <=50K  
...             ...           ...             ...            ...    ...  
30157             0             0              38  United-States  <=50K  
30158             0             0              40  United-States   >50K  
30159             0             0              40  United-States  <=50K  
30160             0             0              20  United-States  <=50K  
30161         15024             0              40  United-States   >50K  

[30162 rows x 14 columns]